scylla/src/observability/metrics.rs

```diff
@@ -1,23 +1,204 @@
 const ORDER_TYPE: Ordering = Ordering::Relaxed;
 
+/// Error that occured upon a metrics operation.
+#[derive(Error, Debug, PartialEq)]
+pub enum MetricsError {
+    #[error("Lock-free histogram error: {0}")]
+    LockFreeHistogramError(#[from] LockFreeHistogramError),
+    #[error("Histogram error: {0}")]
+    HistogramError(#[from] histogram::Error),
+    #[error("Histogram is empty")]
+    Empty,
+}
+
+/// Snapshot is a structure that contains histogram statistics such as
+/// min, max, mean, standard deviation, median, and most common percentiles
+/// collected in a certain moment.
+#[non_exhaustive]
+#[derive(Debug)]
+pub struct Snapshot {
+    pub min: u64,
+    pub max: u64,
+    pub mean: u64,
+    pub stddev: u64,
+    pub median: u64,
+    pub percentile_75: u64,
+    pub percentile_95: u64,
+    pub percentile_98: u64,
+    pub percentile_99: u64,
+    pub percentile_99_9: u64,
+}
+
+/// The interval in seconds for which the rate is calculated.
+const INTERVAL: u64 = 5;
+
+/// An implementation of an Exponentially Weighted Moving Average (EWMA).
 #[derive(Debug)]
-pub struct MetricsError {
-    cause: &'static str,
+struct ExponentiallyWeightedMovingAverage {
+    /// Smoothing factor, a value between 0 and 1.
+    alpha: f64,
+    /// Atomic counter to keep track of the number of requests. \
+    /// Indicates the number of requests that have not been accounted for EWMA yet.
+    uncounted: AtomicU64,
+    /// Atomic boolean to check if the EWMA has been initialized.
+    is_initialized: AtomicBool,
+    ///  Atomic value representing the current rate of requests per second. \
+    ///  AtomicU64 is used to store a floating point number as a bit representation.
+    rate: AtomicU64,
 }
 
-impl From<&'static str> for MetricsError {
-    fn from(err: &'static str) -> MetricsError {
-        MetricsError { cause: err }
+impl ExponentiallyWeightedMovingAverage {
+    fn new(alpha: f64) -> Self {
+        Self {
+            alpha,
+            uncounted: AtomicU64::new(0),
+            is_initialized: AtomicBool::new(false),
+            rate: AtomicU64::new(0),
+        }
+    }
+
+    /// Returns the current `rate` as a floating point number.
+    fn rate(&self) -> f64 {
+        f64::from_bits(self.rate.load(Ordering::Acquire))
+    }
+
+    /// Increments the `uncounted` requests counter. \
+    /// Should be called every time a new request is made.
+    fn update(&self) {
+        self.uncounted.fetch_add(1, ORDER_TYPE);
+    }
+
+    /// Updates the `rate` based on the current number of requests. \
+    /// Should be called every time the interval has passed.
+    ///
+    /// The rate is updated using the formula: \
+    /// `rate = rate + alpha * (instant_rate - rate)` \
+    /// where `instant_rate` is the number of requests in the last interval.
+    ///
+    /// The first time this function is called, the `rate` is set to the `instant_rate`.
+    /// 
+    /// **WARNING: MUST NOT BE CALLED CONCURRENTLY!**
+    fn tick(&self) {
```

Comment:
:thought_balloon: This makes a valid case to employ `unsafe`, I believe, to force callers to prove that the contract is upheld.

Reply:
Will there be a memory corruption if the contract is violated? If not, then it is not a valid use case for unsafe.